src/id_token_provider.erl

```diff
@@ -38,23 +40,57 @@ init([]) ->
   ets:new(?ID_TOKEN_CACHE, ?ETS_OPTIONS),
   Providers = id_token_jwks:get_providers(),
   lists:foreach(fun add_provider/1, Providers),
+  case application:get_env(id_token, async_revalidate, false) of
+    true ->
+        lists:foreach(fun({Provider, _}) ->
+                          self() ! {refresh, Provider}
+                      end, Providers);
+    false -> ok
+  end,
   {ok, #{}}.
 
 handle_call({refresh, Provider}, _From, State) ->
+  {reply, maybe_refresh(Provider), State}.
+
+handle_cast(_Request, State) ->
+  {noreply, State}.
+
+handle_info({refresh, Provider}, State) ->
+  #{exp_at := ExpAt} = refresh(Provider),
+  Now = id_token_util:now_gregorian_seconds(),
+  RevalidateTime = ExpAt - ?REVALIDATE_DELAY,
+  case Now < RevalidateTime of
+    true ->
+      timer:send_after((RevalidateTime - Now) * 1000,
+                       self(), {refresh, Provider});
+    false ->
+      %% Not enough time for revalidation,
+      %% let the first request pay the price
+      ok
```

Comment:
So if the max-age reported by the server is lower than the revalidate time, there's not much that can be done. The async revalidation won't work and the strategy here is to fall back to the default behaviour of the library, which is having the first request that has a cache miss (or a stale value) to revalidate it.

Having the library guessing when it should try again to revalidate is not a good idea in my opinion. So that's why I choose to don't go down that path.

Reply:
Ah I understand, I didn't think about that 'ExpAt' is derived from the cache header. I understand now why you discard the async revalidation strategy completely. And you are right, it does not make sense to guess a time.